cloudinit/distros/alpine.py

```diff
@@ -202,6 +205,374 @@ def preferred_ntp_clients(self):
+                )
+            except IOError as e:
+                util.logexc(LOG, "Failed to update %s file", shadow_file)
+                raise e
+        else:
+            util.logexc(
+                LOG, "Failed to update %s for user %s", shadow_file, name
+            )
+
+    def lock_passwd(self, name):
+        """
+        Lock the password of a user, i.e., disable password logins
+        """
+
+        # Check whether Shadow's or Busybox's version of 'passwd'
+        (_out, err) = subp.subp(["passwd", "--help"])
+        if err != "":
+            first_line = err.splitlines().pop(0)
+            if re.search(r"^BusyBox.*", first_line):
+                # Busybox's 'passwd'
+                use_shadow = False
+            else:
+                # Not Shadow's or Busybox's 'passwd', assume
+                # it functions the same as Shadow's 'passwd'
+                use_shadow = True
+        else:
+            # Shadow's 'passwd --help' does not output anything to stderr
+            use_shadow = True
+
+        # If Shadow's 'passwd' is available then use the generic
+        # lock_passwd function from __init__.py instead.
+        if use_shadow:
+            return super().lock_passwd(name)
+
+        cmd = ["passwd", "-l", name]
+        # Busybox's 'passwd', unlike Shadow's 'passwd', errors
+        # if password is already locked:
+        #
+        #   "passwd: password for user2 is already locked"
+        #
+        # with exit code 1
+        try:
+            (_out, err) = subp.subp(cmd, rcs=[0, 1])
+            if re.search(r"is already locked", err):
+                return True
+        except Exception as e:
+            util.logexc(LOG, "Failed to disable password for user %s", name)
+            raise e
+
+    def expire_passwd(self, user):
+        # Check whether Shadow's or Busybox's version of 'passwd'
+        (_out, err) = subp.subp(["passwd", "--help"])
+        if err != "":
+            first_line = err.splitlines().pop(0)
+            if re.search(r"^BusyBox.*", first_line):
+                # Busybox's 'passwd'
+                use_shadow = False
+            else:
+                # Not Shadow's or Busybox's 'passwd', assume
+                # it functions the same as Shadow's 'passwd'
+                use_shadow = True
+        else:
+            # Shadow's 'passwd --help' does not output anything to stderr
+            use_shadow = True
+
+        # If Shadow's 'passwd' is available then use the generic
+        # expire_passwd function from __init__.py instead.
+        if use_shadow:
+            return super().expire_passwd(user)
+
+        # Busybox's 'passwd' does not provide an expire option
+        # so have to manipulate the shadow file directly.
+        shadow_contents = None
+        shadow_file = self.shadow_fn
+        try:
+            shadow_contents = util.load_text_file(shadow_file)
+        except FileNotFoundError as e:
+            util.logexc(
```

Comment:
I think a `LOG.warning(...)` would suffice here rather than logging the traceback.

Reply:
Done.